src/server/server_family.cc

```diff
@@ -535,23 +535,41 @@ void ClientKill(CmdArgList args, absl::Span<facade::Listener*> listeners, SinkRe
 
   const bool is_admin_request = cntx->conn()->IsPrivileged();
 
+  std::vector<std::vector<facade::Connection::WeakRef>> thread_connections(
+      shard_set->pool()->size());
+
   atomic<uint32_t> killed_connections = 0;
   atomic<uint32_t> kill_errors = 0;
-  auto cb = [&](unsigned thread_index, util::Connection* conn) {
+
+  auto traverse_cb = [&](unsigned idx, util::Connection* conn) {
     facade::Connection* dconn = static_cast<facade::Connection*>(conn);
     if (evaluator(dconn)) {
       if (is_admin_request || !dconn->IsPrivileged()) {
-        dconn->ShutdownSelf();
-        killed_connections.fetch_add(1);
+        thread_connections[idx].push_back(dconn->Borrow());
       } else {
         kill_errors.fetch_add(1);
       }
     }
   };
 
-  for (auto* listener : listeners) {
-    listener->TraverseConnections(cb);
-  }
+  auto cb = [&](unsigned idx, ProactorBase* p) mutable {
+    // Step 1 aggregate the per thread connections from all listeners
```

Comment:
if you do both steps locally in-thread, there is not reason to declare `thread_connections` as a global object outside. Instead just define `vector<facade::Connection::WeakRef>` here

Reply:
Yes I thought about that. My issue was that we need to pass this thread local to `traverse_cb`. This implies either defining the traverse_cb within this lambda which makes it difficult to read or to `bind` it via `[&local] { travere_cb(local, ...); }` 

I am fine with any of these -- I will do whatever you find more readable :)